### PR TITLE
Add support for KNX data secure

### PR DESCRIPTION
**Pull request by nanosonde**
Add support for KNX data secure:
* no access checking with roles is implemented (not planned). The device can just take part in secure communcation. Only the basic Group object security flags checking is implemented for secure group communication.

### CODE REVIEW

**Comment by nanosonde** [comment]
Not tested. Not finished. Just to give an impression of the current status

File: src/knx/security_interface_object.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    };
+    initializeProperties(sizeof(properties), properties);
+}
+
+void SecurityInterfaceObject::secureApplicationLayer(SecureApplicationLayer& secAppLayer)
+{
+    _secAppLayer = &secAppLayer;
+}
+
+uint8_t* SecurityInterfaceObject::save(uint8_t* buffer)
```
Comment (thelsing) [inline src/knx/security_interface_object.cpp:170]: You can just drop the save, restore and saveSize methods.

File: src/knx/application_layer.h [hunk 2]
```diff
@@ -39,12 +40,12 @@ class ApplicationLayer
...
      *        
      * @param apdu The submitted APDU.
-     * 
+     *
      * @param priority The ::Priority of the received request.
      * 
      * @param hopType Should routing be endless or should the NetworkLayer::hopCount be used? See also ::HopCountType.
      */
-    void dataGroupIndication(HopCountType hopType, Priority priority, uint16_t tsap, APDU& apdu);
+    virtual void dataGroupIndication(HopCountType hopType, Priority priority, uint16_t tsap, APDU& apdu) {dataGroupIndication(hopType, priority, tsap, apdu, noSecurity);}
```
Comment (thelsing) [inline src/knx/application_layer.h:48]: I try to keep the header files clean from code.  So these should go to the cpp-files later.

File: src/knx/secure_application_layer.h [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    // to transport layer
+    virtual void dataGroupRequest(AckType ack, HopCountType hopType, Priority priority, uint16_t tsap, APDU& apdu, const SecurityControl& secCtrl) override;
+    virtual void dataBroadcastRequest(AckType ack, HopCountType hopType, Priority priority, APDU& apdu, const SecurityControl& secCtrl) override;
+    virtual void dataSystemBroadcastRequest(AckType ack, HopCountType hopType, Priority priority, APDU& apdu, const SecurityControl& secCtrl) override;
+    virtual void dataIndividualRequest(AckType ack, HopCountType hopType, Priority priority, uint16_t destination, APDU& apdu, const SecurityControl& secCtrl) override;
+    virtual void dataConnectedRequest(uint16_t tsap, Priority priority, APDU& apdu, const SecurityControl& secCtrl) override; // apdu must be valid until it was confirmed
+
+  private:
+
+    template <typename K, typename V, int SIZE>
```
Comment (thelsing) [inline src/knx/secure_application_layer.h:62]: We should probably put this template in its own file.

File: src/knx/security_interface_object.cpp (hunk 1 above)
Comment (nanosonde) [reply src/knx/security_interface_object.cpp:171]: Yes, I will remove it if I clean up everything.

File: src/knx/application_layer.h (hunk 2 above)
Comment (nanosonde) [reply src/knx/application_layer.h:48]: Ok, will move it.

File: src/knx/secure_application_layer.h (hunk 3 above)
Comment (nanosonde) [reply src/knx/secure_application_layer.h:63]: yep, will move it "SimpleMap.h".

File: src/knx/interface_object.h [hunk 4]
```diff
@@ -125,7 +128,7 @@ class InterfaceObject : public SaveRestore
      *
      * @param[out] resultData The result data for the function
      */
-    virtual void state(PropertyID id, uint8_t* data, uint8_t length, uint8_t* resultData, uint8_t resultLength);
+    virtual void state(PropertyID id, uint8_t* data, uint8_t length, uint8_t* resultData, uint8_t &resultLength);
```
Comment (nanosonde) [inline src/knx/interface_object.h:131]: Bugfix.

File: src/knx/secure_application_layer.h (hunk 3 above)
Comment (nanosonde) [reply src/knx/secure_application_layer.h]: done.

File: src/knx/application_layer.h (hunk 2 above)
Comment (nanosonde) [reply src/knx/application_layer.h]: done

File: src/knx/security_interface_object.cpp (hunk 1 above)
Comment (nanosonde) [reply src/knx/security_interface_object.cpp]: done.

**Comment by thelsing** [comment]
Can you resolve the conflicts with the main branch please?

**Comment by nanosonde** [comment]
So, now let's get it merged. 

  | :grin:
-- | --

**Pull request merged**